machineconfiguration/v1alpha1/types_machineosbuild.go

```diff
@@ -0,0 +1,174 @@
+	// +listType=set
+	// +optional
+	BuildHistory []PriorMachineOSBuilds `json:"buildHistory,omitempty"`
+}
+
+// BuildProgess highlights some of the key phases of a build to be tracked in Conditions.
+type BuildProgress string
+
+const (
+	// prepared indicates that the build has finished preparing. A build is prepared
+	// by gathering the build inputs, validating them, and making sure we can do an update as specified.
+	MachineOSBuildPrepared BuildProgress = "Prepared"
+	// building indicates that the build has been kicked off with the specified image builder
+	MachineOSBuiling BuildProgress = "Building"
+	// failed indicates that during the build or preparation process, the build failed.
+	MachineOSBuildFailed BuildProgress = "Failed"
+	// interrupted indicates that the user stopped the build process by modifying part of the build config
+	MachineOSBuildInterrupted BuildProgress = "Interrupted"
+	// restarted indicates that this build has been either interrupted or failed, and now a new
+	// iteration of the process has begun.
+	MachineOSBuildRestarted BuildProgress = "Restarted"
+	// ready indicates that the build has completed and the image is ready to roll out.
+	MachineOSReady BuildProgress = "Ready"
+)
+
+// PriorMachineOSBuilds contains information about related builds
+type PriorMachineOSBuilds struct {
+	// name is the name of the build
+	// +kubebuilder:validation:Required
+	Name string `json:"name"`
+	// buildFailure contains an optional message of why this build ended prematurely.
+	// +optional
+	BuildFailure string `json:"buildFailure"`
+}
+
+// Refers to the name of a MachineConfigPool (e.g., "worker", "infra", etc.):
+type MachineConfigPoolReference struct {
+	// name is the name of the referenced object.
+	// +kubebuilder:validation:Required
+	Name string `json:"name"`
+}
+
+// Refers to the name of a rendered MachineConfig (e.g., "rendered-worker-ec40d2965ff81bce7cd7a7e82a680739", etc.):
+type RenderedMachineConfigReference struct {
+	// name is the name of the referenced object.
+	// +kubebuilder:validation:Required
+	Name string `json:"name"`
+}
+
+// Refers to the name of a (future) MachineOSImage (e.g., "worker-os-image-167651b10ec98af17971d6a47df9e22f", etc.):
+type MachineOSImageReference struct {
+	// name is the name of the referenced object.
+	// +kubebuilder:validation:Required
+	Name string `json:"name"`
+}
+
+// Refers to the name of an image registry push/pull secret needed in the build process.
+type ImageSecretObjectReference struct {
+	// name is the name of the referenced object.
+	// +kubebuilder:validation:Required
+	Name string `json:"name"`
+}
+
+// BuildInputs holds all of the information needed to trigger a build
+type BuildInputs struct {
+	// imageBuilderType specifies the backend to be used to build the image.
+	// Valid options are: OpenShiftImageBuilder, PodImageBuilder, and Default (OpenShiftImageBuilder)
+	ImageBuilderType MachineOSImageBuilderType `json:"imageBuilderType"`
```

Comment:
note: Right now, I don't think there is a way to bind an image builder type to a specific pool. This feels like it should be a global option in the future.

Reply:
this can be MCO validation that we run I believe.